security/pkg/workload/server.go

```diff
@@ -190,16 +180,34 @@ func (s *SDSServer) RegisterUdsPath(udsPath string) error {
 		return fmt.Errorf("failed to listen on %v", err)
 	}
 
+	var opts []grpc.ServerOption
+	udsServer := grpc.NewServer(opts...)
+	sds.RegisterSecretDiscoveryServiceServer(udsServer, s)
+	s.udsServerMap[udsPath] = udsServer
+
 	// grpcServer.Serve() is a blocking call, so run it in a goroutine.
 	go func() {
-		log.Infof("Starting GRPC server on UDS path", udsPath)
-
-		err := s.grpcServer.Serve(listener)
+		log.Infof("Starting GRPC server on UDS path: %s", udsPath)
+		err := udsServer.Serve(listener)
 		// grpcServer.Serve() always returns a non-nil error.
 		log.Warnf("GRPC server returns an error: %v", err)
 	}()
 
 	return nil
 }
 
-// TODO: add methods to unregister the deleted UDS paths and the listeners.
+// DeregisterUdsPath closes and removes the grpcServer instance serving UDS
+func (s *SDSServer) DeregisterUdsPath(udsPath string) error {
+	s.udsServerMapGuard.Lock()
+	defer s.udsServerMapGuard.Unlock()
+
+	if udsServer, ok := s.udsServerMap[udsPath]; ok {
```

Comment:
I feel the logic can be made more readable :)

if !ok {
   return error...
}
everything else
return nil

Reply:
Done